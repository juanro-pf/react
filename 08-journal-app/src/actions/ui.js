import { types } from "../types/types";

export const setError= (err) => ({
  type: types.uiSetError,
  payload: err
});

export const UnsetError= () => ({
  type: types.uiUnsetError
});

export const startLoading= () => ({
  type: types.uiStartLoading
});

export const finishLoading= () => ({
  type: types.uiFinishLoading
});