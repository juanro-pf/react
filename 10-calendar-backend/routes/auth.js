/*
  Rutas de Auth
  host + /api/auth
*/

const { Router }= require('express');
const { check }= require('express-validator');
const { validateFields } = require('../middlewares/fieldValidator');
const { createUser, loginUser, revalidateToken } = require('../controllers/auth');
const { validateJWT }= require('../middlewares/validateJWT');

const router= Router();

router.post(
  '/new',
  [ //middlewares
    check('name', 'Name is required').not().isEmpty(),
    check('email', 'Email is required').isEmail(),
    check('password', 'Password must contain 6 characters').isLength({ min: 6 }),
    validateFields
  ],
  createUser
);

router.post(
  '/',
  [
    check('email', 'Email is required').isEmail(),
    check('password', 'Password must contain 6 characters').isLength({ min: 6 }),
    validateFields
  ],
  loginUser
);

router.get(
  '/renew',
  validateJWT,
  revalidateToken
);

module.exports= router;