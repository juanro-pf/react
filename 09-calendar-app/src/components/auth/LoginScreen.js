import React from 'react';
import './login.css';
import { useForm } from '../../hooks/useForm';
import { useDispatch } from 'react-redux';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

export const LoginScreen = () => {

  const dispatch = useDispatch();

  const [ formLoginValues, handleLoginInputChange ]= useForm({
    lEmail: 'jro@gmail.com',
    lPassword: '123456'
  });

  const [ formRegisterValues, handleRegisterInputChange ]= useForm({
    rName: 'Ropo',
    rEmail: 'jropo@gmail.com',
    rPassword: '123456',
    rPassword2: '123456'
  });

  const { lEmail, lPassword }= formLoginValues;
  const { rEmail, rName , rPassword , rPassword2 }= formRegisterValues;

  const handleLogin= (e) => {
    e.preventDefault();
    dispatch( startLogin( lEmail, lPassword ) );
  };

  const handleRegister= (e) => {
    e.preventDefault();

    if( rPassword !== rPassword2 ){
      return Swal.fire('Error', 'Passwords must match', 'error');
    }
    
    dispatch( startRegister( rEmail, rName, rPassword ) );
  };

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Log in</h3>
          <form onSubmit= { handleLogin }>
            <div className="form-group">
              <input 
                type="text"
                className="form-control"
                placeholder="Email"
                name= 'lEmail'
                value= { lEmail }
                onChange= { handleLoginInputChange }
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password"
                name= 'lPassword'
                value= { lPassword }
                onChange= { handleLoginInputChange }
              />
            </div>
            <div className="form-group">
              <input 
                type="submit"
                className="btnSubmit"
                value="Login" 
              />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Register</h3>
          <form onSubmit= { handleRegister }>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Name"
                name= 'rName'
                value= { rName }
                onChange= { handleRegisterInputChange }
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Email"
                name= 'rEmail'
                value= { rEmail }
                onChange= { handleRegisterInputChange }
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password" 
                name= 'rPassword'
                value= { rPassword }
                onChange= { handleRegisterInputChange }
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repeat password"
                name= 'rPassword2'
                value= { rPassword2 }
                onChange= { handleRegisterInputChange }
              />
            </div>

            <div className="form-group">
              <input 
                type="submit" 
                className="btnSubmit" 
                value="Create account"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  )
}