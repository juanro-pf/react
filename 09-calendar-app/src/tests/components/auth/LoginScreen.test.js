import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';
import Swal from 'sweetalert2';

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));

jest.mock('../../../actions/auth', () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn()
}));

const middlewares= [thunk];
const mockStore= configureStore( middlewares );

const initState= {};
const store= mockStore(initState);
store.dispatch= jest.fn();

const wrapper= mount(
  <Provider store= {store}>
    <LoginScreen />
  </Provider>
);

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Testing LoginScreen', () => {
  
  test('should match snapshot', () => {
    
    expect(wrapper).toMatchSnapshot();
  });
  
  test('should call dispach(login())', () => {
    
    wrapper.find('input[name="lEmail"]').simulate('change', {
      target: {
        name: 'lEmail',
        value: 'juan@gmail.com'
      }
    });

    wrapper.find('input[name="lPassword"]').simulate('change', {
      target: {
        name: 'lPassword',
        value: '123456'
      }
    });

    wrapper.find('form').at(0).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startLogin).toHaveBeenCalledWith("juan@gmail.com", "123456");
  });

  test('should fail register if passwords dont match', () => {
    
    wrapper.find('input[name="rPassword"]').simulate('change', {
      target: {
        name: 'rPassword',
        value: '123456'
      }
    });

    wrapper.find('input[name="rPassword2"]').simulate('change', {
      target: {
        name: 'rPassword2',
        value: '1234567'
      }
    });

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startRegister).not.toHaveBeenCalled();
    expect(Swal.fire).toHaveBeenCalledWith('Error', 'Passwords must match', 'error');

  });
  
  test('should call register', () => {
    
    wrapper.find('input[name="rPassword"]').simulate('change', {
      target: {
        name: 'rPassword',
        value: '123456'
      }
    });

    wrapper.find('input[name="rPassword2"]').simulate('change', {
      target: {
        name: 'rPassword2',
        value: '123456'
      }
    });

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startRegister).toHaveBeenCalled();
    expect(Swal.fire).not.toHaveBeenCalled();

  });

});
