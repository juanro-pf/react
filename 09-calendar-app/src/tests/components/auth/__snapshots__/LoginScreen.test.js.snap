// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing LoginScreen should match snapshot 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Log in
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value="jro@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value="Ropo"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value="jropo@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repeat password"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Create account"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
